src/si/time.rs

```diff
@@ -55,17 +55,18 @@ where
     nanosecond: ::Conversion<V, T = V::T>,
 {
     fn from(t: Time<U, V>) -> Duration {
-        let mut secs = t.get::<second>();
-        if secs < V::zero() {
-            secs = V::zero() - secs;
-        }
-        let secs = secs.as_();
+        let t = if t < Time::<U, V>::new::<second>(V::zero()) {
```

Comment:
I understand why this is here, but do we want to silently convert negative times to (inherently positive) durations? I suppose this might be a matter of "time vs. time interval."

Reply:
I don't think it's really a silent conversion, the user has to manually call `into` in a context where the conversion is into an inherently positive type.  I guess it's a question of which is least surprising:

1) Absolute Value
2) Wrapping Around
3) Minimum 0

Reply:
I know that I would prefer panicking, although I’d imagine that others would see it differently.

Reply:
Really?  I just put that option there for completeness, From [**must not fail**](https://doc.rust-lang.org/std/convert/trait.From.html)

Reply:
In the context of that documentation page, "fallible functions" seem to be loosely be "functions that return `Result<T, E>` or `Option<T>`:

> When constructing a function that is capable of failing the return type will generally be of the form `Result<T, E>`.

As such, I read that note as simply meaning "`From` does not return an error type."  A more general argument for this interpretation is that it's almost nonsensical to say "this code must not panic" (with a couple of notable exceptions) in Safe Rust.

Panics signal an irrecoverable error, something that makes no sense and for which there's no sensible alternative route. In my mind, converting a negative time to a `Duration` is such an error, and it makes sense to panic, as this likely signals a logic error. However, panicking from within a library crate is certainly not ideal.

I actually might argue, if possible, that we should use `TryFrom` instead of `From`.

Reply:
So `TryFrom` vs `From` is a different issue, there's certainly an argument to be made there.  The trick is, doing that sort of error management (while certainly more rigorous) is enough of an ergonomics hit that if that's the direction `uom` decides to go I'm going to just continue defining my own conversion functions where and as I need them.  I'm not saying you're wrong, you're not, `TryFrom` is definitely a more correct fit, it just feels cumbersome (and the implementation is going to be even more of a nightmare).

So with that out of the way, I think your reading of the note is wrong, and here's why.  There's absolutely no point saying "`From` does not return an error type", because there is no flexibility in `from`'s return type.  There's simply no choice what you return.  It would be warning us to make sure our code builds, or warning us to not write bugs into our code.

If you assume they added (and emphasized) that warning for a reason, there must be more to it than that.  It must be more general than "don't do what you can't do".  The reason they warn us not to let it fail is because From is integral to Rust's failure handling mechanisms, and it's not at all clear what should happen when failure handling fails.

Reply:
... so... I think I just accidentally convinced myself that this PR should be rejected.  *sigh*

Reply:
Not to beat a dead horse, but I still believe that the meaning for that documentation for `From` is simply "the return value is not an error type." This is especially supported by the fact that [you can `impl From<A> for Option<B>` to mimic `TryFrom`](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=c1b0510ed7f7d091e19ab12ff7d8564e) (and so there is indeed some flexibility in the return type). Regardless, I do think that `TryFrom` is the better fit here. May I ask why `From` is necessary for your use case/why `TryFrom` is insufficient?

Reply:
It's not that the trait is insufficient on its own (I mean, it's a bit cumbersome to have to do yet more error handling, but at the end of the day that's not that bid a deal), it's more that the difficulty of making the change is greater than the amount I care about it being made.  I've already sunk 100x more time into this PR than its actually worth to me, and if I were to double that investment I still don't think that's enough to adequately implement TryFrom.

Reply:
That makes sense, thanks. I think I might try to make up a `TryFrom` PR, fwiw. Thanks for your efforts on this.